현재 개선해야 하는 부분

InJun2 opened the pull request:
- switch 를 이용한 로직 비교 -> 이걸 어떤 방식으로 바꿀 수 있는지 고민
- if 문을 interface 로 변경하는 방법 고민
- compareRepeatedResult 에 로직이 너무 김 -> 줄이기 목표

InJun2 committed:
- input, output 객체를 하나씩 만들어서 main 에서 입력/출력
- input 은 기본적으로 정수여야 하고, 시뮬레이션 시작은 0,1만 입력가능, 시뮬레이션 횟수는 0~100회 까지만 가능
- 랜덤 당첨 번호 6개와 보너스번호 1개 생성
- 반복할 시뮬레이션 횟수를 입력받으면 해당 횟수만큼 자동으로 6개의 랜덤 번호 생성
- 이후 해당 시뮬레이션이 담긴 리스트를 리스트 크기만큼 당첨번호와 비교하여 ComparisonNumbers 클래스에 simulationResults 값을 넣을 예정
- 이후 해당 값을 Get 메소드로 리턴받아 main 에서 output으로 출력 예정

InJun2 committed:
- switch 를 이용한 로직 비교 -> 이걸 어떤 방식으로 바꿀 수 있는지 고민
- if 문을 interface 로 변경하는 방법 고민
- compareRepeatedResult 에 로직이 너무 김 -> 줄이기 목표

DongGeon0908 commented on src/main/java/view/OutputScreen.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public void outputSimulatorResult(SimulatedLotto simulatedLotto, ComparisonNumbers comparisonNumbers) {
+        int i = 0;
+        System.out.println("********* 자동 생성 번호 *********");
+        for (SimulatedLottoNumber SimulatedLottoNumber : simulatedLotto.getSimulationLotto()) {
+            for (int number : SimulatedLottoNumber.getNumbers()) {
+                System.out.print(number + " ");
+            }
+            System.out.println("-> " + comparisonNumbers.getSimulationLottoResult().get(i));
+            i++;
+        }
+        System.out.println("*******************************");
```
너무 많은 로직이 메서드에 있네요
for문 2번쓰는건 지양해주세요!

DongGeon0908 commented on src/main/java/run/ComparisonNumbers.java:
```diff
@@ -0,0 +1,55 @@
+        return simulationResults;
+    }
+
+    private String toStringResult(int winningNumber, boolean bonusWinning) {
+        switch (winningNumber) {
+            case 6:
+                return "1등";
+            case 5:
+                if (bonusWinning) {
+                    return "2등";
+                }
+                return "3등";
+            case 4:
+                return "4등";
+            case 3:
+                return "5등";
+            default:
+                return "낙첨";
+        }
+    }
```
enum으로 구현해보시죠!

DongGeon0908 commented on src/main/java/run/ComparisonNumbers.java:
```diff
@@ -0,0 +1,55 @@
+        this.simulationResults = compareRepeatedResult(winningNumber, simulationLotto);
+    }
+
+    private List<String> compareRepeatedResult(WinningNumber winningNumber, SimulatedLotto simulatedLotto) {
+        List<String> simulationResults = new ArrayList<>();
+
+        for (int i = 0; i < simulatedLotto.getSimulationLotto().size(); i++) {
+            int winningResult = 0;
+            boolean bonusWinning = simulatedLotto.getSimulationLotto().get(i).getNumbers().contains(winningNumber.getBonusNumber());
+            for (Integer integer : simulatedLotto.getSimulationLotto().get(i).getNumbers()) {
+                if (winningNumber.getNumbers().contains(integer)) {
+                    winningResult++;
+                }
+            }
+
+            simulationResults.add(toStringResult(winningResult, bonusWinning));
+        }
+        return simulationResults;
+    }
```
메서드가 길어진다면, 해당 메서드가 너무 많은 역할과 책임을 갖고 있다는 것을 증명하는 것입니다!!
조금 더 쪼개보시죵

DongGeon0908 commented on src/main/java/run/ComparisonNumbers.java:
```diff
@@ -0,0 +1,55 @@
+public class ComparisonNumbers {
+    private final List<String> simulationResults;
+
+    public ComparisonNumbers(WinningNumber winningNumber, SimulatedLotto simulationLotto) {
+        this.simulationResults = compareRepeatedResult(winningNumber, simulationLotto);
+    }
```
생성자를 이용하는 건 굳굳

DongGeon0908 commented on src/main/java/model/SimulatedLottoNumber.java:
```diff
@@ -0,0 +1,24 @@
+        this.simulationNumbers = generateSimulationNumbers();
+    }
+
+    private Set<Integer> generateSimulationNumbers() {
```
stream을 써보는 건 어떨까요?

DongGeon0908 reviewed (changes requested):
굳굳입니다!!
아래 내용을 참고해서 수정해보시면 좋을 것 같아요!

1. 역할과 책임을 생각하며 메서드 쪼개기!
2. enum 활용하기!
3. stream 써보기!

InJun2 committed:
- WinningResult enum 생성
- switch 를 이용한 로직 비교 -> enum 값을 stream 으로 변환하여 탐색
- SimulatedLottoNumber 랜덤 숫자 생성을 stream 으로 생성
- compareRepeatedResult 에 로직이 너무 김 -> 메소드 분산

DongGeon0908 commented on src/test/java/ExampleTest.java:
```diff
@@ -1,34 +1,39 @@
 /**
  * 확인 후 해당 파일을 삭제하시기 바랍니다.
  */
 public class ExampleTest {
```
이번 기회에 테스트 코드를 작성하시는 건 어떨까요?

DongGeon0908 commented on src/main/java/view/OutputScreen.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class OutputScreen {
```
```suggestion
public class OutputView {
```

DongGeon0908 commented on src/main/java/view/OutputScreen.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void outputSimulatorResult(SimulatedLotto simulatedLotto, ComparisonNumbers comparisonNumbers) {
+        AtomicInteger counter = new AtomicInteger(0);
```
AtomicInteger 요것이 무엇일까요?

InJun2 replied:
int i를 지역변수로 사용한 후 람다식 반환값에 i++; 를 사용했더니 오류가 발생했고 자동 해결을 통해 AtomicInteger로 변환되었음.
해당 클래스가 무엇인지 모르고 사용했는데 이후 찾아보니 주로 동기화를 위해 사용한다고 함
멀티스레드 환경에서 동시성 관리를 위해 도입된 자료형으로 스트림을 순회하면서 외부 변수에 접근하려고 하면 동시성 관리를 위해 Atomic 자료형을 사용해야한다고 함

DongGeon0908 commented on src/main/java/run/ComparisonNumbers.java:
```diff
@@ -0,0 +1,50 @@
+        this.simulationResults = compareRepeatedResult(winningNumber, simulationLotto);
+    }
+
+    private List<String> compareRepeatedResult(WinningNumber winningNumber, SimulatedLotto simulatedLotto) {
+        List<String> simulationResults = new ArrayList<>();
+
+        for (int i = 0; i < simulatedLotto.getSimulationLotto().size(); i++) {
+            List<Integer> numbers = simulatedLotto.getSimulationLotto().get(i).getNumbers();
+
+            simulationResults.add(toStringResult(getWinningResult(numbers, winningNumber), getBonusWinningResult(numbers, winningNumber)));
+        }
+        return simulationResults;
+    }
+
+    private int getWinningResult(List<Integer> simulatedLottoNumbers, WinningNumber winningNumber) {
+        int winningResult = 0;
+        for (Integer lottoNumber : simulatedLottoNumbers) {
+            if (winningNumber.getNumbers().contains(lottoNumber)) {
+                winningResult++;
+            }
+        }
+        return winningResult;
+    }
```
stream으로 돌리는 것도 좋을 것 같네요

DongGeon0908 reviewed (changes requested):
좋은 것 같아요!!
이번에 목표는 테스트 코드입니다!!

다음 리뷰 요청에서는 테스트 코드를 작성해주셔서 리뷰 요청 부탁드립니다.
(추가적으로 코드를 조금 더 예쁘게?? 작성하는 연습도!!)

InJun2 committed:
- LottoSimulator 에서 상수 public 으로 변경
- ComparisonNumbers 메소드를 stream 으로 변경

InJun2 committed:
- 테스트코드 추가

DongGeon0908 reviewed (commented):
고생했어요
별도로 리뷰드릴 내용은 없을 것 같네요
현재 코드를 검증할 수 있는 테스트 코드를 작성해주세요

InJun2 committed:
- 테스트코드 추가
